### ISSUE TITLE
Allow effects to influence vitamins

### ISSUE THREAD

**Issue by Venera3** [CleverRaven/Cataclysm-DDA#46947]
# Is your feature request related to a problem? Please describe.

Effects are nice and flexible, but unlike muations they aren't hooked up to the vitamin system. Adding that functionality would allow for realistic malnutrition effects and would provide a potential long-term scaffolding for more complicated effects.

# Describe the solution you'd like
Adding base and scaling modifiers to modify the natural decay rate and absorption of a given vitamin, to the tune of:

```
"vitamin_rate": [ { "id": "vitc", -10, -5 } ],
"vitamin_absorb_multi": [ { "id": "vitc", 0.2, 0.5 } ], 
"vitamin_rate_tick": [ 5, 10 ],
```
with the first value being applied to a susceptible character and the second to a resistant one, as with the other effect modifiers.

# Describe alternatives you've considered

Hardcoding modifiers to the existing vitamins with
``` 
"vitc_mod": [ -10, -5 ],
"iron_mod": [ -10, -5 ],
etc.
```
probably needs less structural changes to the effect code, and would be a welcome addition on its own.

# Additional context
Some use cases:
- Iron deficiency slowing the regeneration of red blood cells, and blood loss/anemia increasing the required iron consumption in exchange for faster blood cell regeneration,
- Monster attacks/spells interfering with vitamins: Mosquitos sucking blood, the venom of heavily mutated creatures raising Toxins,
- Long-term effects of repeated exposure: repeated smoke inhalation leading to progressive stamina maluses,
- If the generic solution were implemented that would open the door to sequential effects, or effects with repeated exacerbations: Monster/drug A applies a permanent effect A, raising a custom vitamin with a set rate. After the vitamin reaches its deficiency threshold it applies Effect B, a self-intensifying effect that lowers the vitamin below its trigger after reaching a set intensity. Magiclysm witches cursing you with recurring Tindalos rifts you need to find a hidden altar to permanently cleanse, anyone?

**Comment by Venera3** [comment]
@anothersimulacrum , this with `vitamin_rate` turned probabilistic ( `"vit_rate_min", "vit_rate_max", "vit_rate_chance"`) and a `"chance_kill"` field would also let me JSON both infections and blood loss. For the wound infection, the sequence would be:

- `bite` stays unchanged, progresses in 6 hours to `local_infection` and can be disinfected away
- `local_infection` is a *permanent* effect, having pain and some limb dmg as its own effect and ticking up a vitamin component (something creative like `"infectvit"`) randomly, with the average rate being positive for both normal and `INFRESIST` characters. It can be removed by proper wound cleaning, requiring cutting, disinfectant, bandages (maybe a way to check for sterility), healthcare of 2-3, causing limb dmg, pain and having a chance to fail or an amputation with lower requirements
- `"infectvit"` has a baseline decay, higher for `INFRESIST` characters but lower than the average of gaining the vitamin from a `local_infection`. It has 3/4 deficiency intensities, causing the systemic debuffs of sepsis/SIRS equivalent to the current infection, with the last intensity activating `"chance_kill"`, a random chance to kill you
- Antibiotics become a common effect, with the different drugs applying different intensities and the effect itself having `"int_add_val": 0` to prevent them stacking. They modify the `"infectvit"` decay rate, with weak counteracting the gain but not lowering it on average, and the stronger variants actually lowering it. Panaceus becomes a much stronger, separate effect to allow stacking both

This is obviously my greediest pitch and the way we have the most generic solution for other effects to get in on the fun, and is based on my complete lack of understanding of what's doable with the effect code.

**Comment by anothersimulacrum** [comment]
So, request is:

- [ ] effects can modify vitamin levels (by different amounts if resistant or not)
- [ ] effects can modify dietary absorption of vitamins (or all sources?) (by different amounts if resistant or not)
- [ ] how often the effect modifies vitamin levels can be specified (with different levels for if resistant or not)
- [ ] effects can have a chance to kill the player specified
- [ ] vitamins can have randomness introduced into their decay/generation
- [ ] sort out all the hardcoded bits with infections to work as you'd like (no new infrastructure, just changes in how we do things)

Am I missing anything, is anything incorrect?

**Comment by actual-nh** [comment]
- Have it possible for some of the randomness to be influenced by health.
- While this would not be part of this proposal itself (I would think), making sure to not get in the way of traits eventually being able to alter _starting_ vitamin levels would be nice.

**Comment by Venera3** [comment]
> Am I missing anything, is anything incorrect?

No, that's all of it. The general randomness/tick part feels like it could use much of the existing effect plumbing, ie:
```
X_min: [ A, B ],  // lower bound, B for resistant chars
X_max: [ A, B],  // upper bound, B for resist
X_tick: [ A, B], 
etc.
```
A generic way of declaring vitamins  ie. having them modify freely-defined vitamins instead of hardcoding a separate modifier for each vitamin would be a large deviation from the code in place, but it is not imperative (but then no self-cursing curses for lizardmen).

On the non-effect front we'd need to 
- change the `bite` part to apply `local_infection` at a defined intensity instead of duration (intensity can be JSON-adjusted and can be set to accumulate in defined steps, so that makes tuning easier) 
- change the bite attacks to apply the full duration and repead bites to add intensity instead
- add an action for wound cleaning with failure states
- add the same to Autodocs/surgeons

> influenced by health

I'm very much not a fan of what health tries to represent or the way it works (I haven't dove into it, but I get the impression that it's prone to negative feedback loops), so I wouldn't be enthusiastic about involving it. Also, the point of hooking the long-term component up to the vitamin system would be to drag most of the effect into JSON allowing for easy adjustments and more transparence, and health helps with neither of those things.

**Comment by actual-nh** [comment]
> > influenced by health
> 
> I'm very much not a fan of what health tries to represent or the way it works (I haven't dove into it, but I get the impression that it's prone to negative feedback loops), so I wouldn't be enthusiastic about involving it. Also, the point of hooking the long-term component up to the vitamin system would be to drag most of the effect into JSON allowing for easy adjustments and more transparence, and health helps with neither of those things.

No problem. (I personally would split "health" into multiple components, including one/two for immune system function - two because of many ways to boost this likely boosting unwanted parts like allergies and autoimmune disease (the ROT series of traits, perhaps?).) I was going off of the infection discussion and your mentioning that the recovery rolls currently are influenced by health, as I understood it.

**Comment by actual-nh** [comment]
Note, BTW, that there should be some chance of a local infection not going systemic - a small chance, but one that's there - but we can work that out once the "knobs" are in JSON. (I think that, if it goes to the level where amputation is absolutely needed, a systemic infection is a reasonable substitute, unless someone wants to code up amputations...)

**Comment by Venera3** [comment]
As long as the lower bound of the vitamin increase is below the natural decay, you have a low chance of not progressing, ex.
``` 
vit_min= 5
vit_max = 25 
vit_tick = 30 s
decay_rate= 3 s
```
comes out to average +5 per tick (assuming 100% chance), but has the possibility of coming out as low as -5. This would also have the fun consequence that a local infection would be able to be kept in check (still causing pain etc, just not progressing) by antibiotics. The local process itself could have intensity levels, with increasing vitamin production, meaning that while weak antibiotics might halt it at first, if you slip up / run out they will not help you anymore.

Honestly, untreated wounds that are fitting the game's context are ( thankfully)  not a field I have much experience in. The clientele that fits most survivors -young, no underlying conditions, acute (poly)trauma- gets brought in and treated promptly, with days/weeks-old wounds being typical for those with severe chronic conditions. Most really long-term wound infections I have the misfortune of meeting are secondary infections after the primary wound care and usually require repeat operations to stay local (but that's the point where my selection bias creeps in, admittedly).

**Comment by actual-nh** [comment]
> Honestly, untreated wounds that are fitting the game's context are ( thankfully) not a field I have much experience in. The clientele that fits most survivors -young, no underlying conditions, acute (poly)trauma- gets brought in and treated promptly, with days/weeks-old wounds being typical for those with severe chronic conditions. Most really long-term wound infections I have the misfortune of meeting are secondary infections after the primary wound care and usually require repeat operations to stay local (but that's the point where my selection bias creeps in, admittedly).

I have encountered (as a beggar) a person who had refused amputation and left the hospital with a wound infection. One could see the bone through the necrotic tissue... shudder. Thanks to the nerves being dead, he was still hobbling on that leg...

I note that there are times where your long-term secondary infection experience would apply - primary wound care done but not quite successfully (by an amateur - more than likely - attempting self-surgery).

**Comment by actual-nh** [comment]
> As long as the lower bound of the vitamin increase is below the natural decay, you have a low chance of not progressing, ex.
> 
> ```
> vit_min= 5
> vit_max = 25 
> vit_tick = 30 s
> decay_rate= 3 s
> ```
> 
> comes out to average +5 per tick (assuming 100% chance), but has the possibility of coming out as low as -5. This would also have the fun consequence that a local infection would be able to be kept in check (still causing pain etc, just not progressing) by antibiotics. The local process itself could have intensity levels, with increasing vitamin production, meaning that while weak antibiotics might halt it at first, if you slip up / run out they will not help you anymore.

A nice simulation of antibiotic resistance!
